Also:
```rust
92.20 error[E0061]: this function takes 24 arguments but 25 arguments were supplied
92.20    --> src/api.rs:101:23
92.20     |
92.20 101 |         let econfig = EngineConfig::new(
92.20     |                       ^^^^^^^^^^^^^^^^^
92.20 ...
92.20 126 |             None, // force_cache - not exposed in simple API
92.20     |             ---- unexpected argument #25 of type `std::option::Option<_>`
92.20     |
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "vllm-rs"
-version = "0.6.12"
+version = "0.6.13"
 edition = "2021"
 default-run = "vllm-rs"
 

--- a/ReadMe-CN.md
+++ b/ReadMe-CN.md
@@ -98,11 +98,11 @@ python3 -m pip install vllm_rs
 ```
 
 ### 🌐✨ API Server + ChatGPT风格内置网页
-   💡你可以使用**任何兼容 OpenAI API 的客户端**进行交互
-   
    💡使用`--ui-server`会同时启动ChatGPT风格网页, 此时无需其它客户端。
 
    💡如长文本请求导致当前生成过程卡顿，请使用 **Rust PD Server**方案 （见**PD分离**）
+
+   💡当`context-cache`启用时，使用`--force-cache` 强制启用fingerprint对话检测，此时客户端无需提供 `session_id`
 
    ⚠️ 过度量化可能会在推理模型中触发 **“思考过程被截断”** 问题。建议采用 BF16 或 Q6K/Q8_0；GPTQ 或 AWQ 也可用于缓解该问题，也可以关闭`context-cache`或 通过`thinking=False` / `enable_thinking=False`关闭推理过程。
 
@@ -350,6 +350,7 @@ pip install target/wheels/vllm_rs-*-cp38-abi3-*.whl --force-reinstall
 | `--ui-server`       |  服务模式: 启动API服务，同时启动ChatGPT风格的内置对话网页服务 |
 | `--kv-fraction`       |  用于控制KVCache使用量 (模型加载后剩余可用GPU显存的百分比) |
 | `--context-cache`   | 启用上下文缓存，用于多轮对话 |
+| `--foce-cache`       | 启用Fingerprint对话标识检测，当 `--context-cache`同时启用时，自动进行上下文缓存，无需客户端显示传入 `session_id`  |
 
 ### MCP配置参数
 

--- a/ReadMe.md
+++ b/ReadMe.md
@@ -102,11 +102,11 @@ python3 -m pip install vllm_rs
 
 ### 🌐✨ API Server + Built-in ChatGPT-like Web Server
 
-💡You can use **any client compatible with the OpenAI API** to interact
-
 💡Start with `--ui-server` will also start ChatGPT-like web server, no external chat client required in that case.
 
 💡Use the Rust PD Server (see **PD Disaggregation**) if decoding stalls during prefilling of long-context requests.
+
+💡When `context-cache` enabled, `--force-cache` enables fingerprint-based session detection if `session_id` not provided in the client side.
 
  ⚠️Low quantization may cause **"Thinking Process Truncated"** for reasoning models, use BF16, Q6K/Q8_0, GPTQ/AWQ can mitigate the problem, or disable `context-cache` or disable reasoning through `thinking=False` / `enable_thinking=False` might also help.
 
@@ -115,7 +115,7 @@ python3 -m pip install vllm_rs
 
 ```bash
 # CUDA
-python3 -m vllm_rs.server --m unsloth/Qwen3-30B-A3B-Instruct-2507-GGUF --f Qwen3-30B-A3B-Instruct-2507-Q4_K_M.gguf --kv-fraction 0.6 --ui-server --context-cache
+python3 -m vllm_rs.server --m unsloth/Qwen3-30B-A3B-Instruct-2507-GGUF --f Qwen3-30B-A3B-Instruct-2507-Q4_K_M.gguf --kv-fraction 0.6 --ui-server --context-cache --force-cache
 # Metal/MacOS (response can be seriously degradated on MacOS pre-Tahoe, use a smaller `--max-model-len` or `--kv-fraction` parameter)
 python3 -m vllm_rs.server --m unsloth/Qwen3-4B-GGUF --f Qwen3-4B-Q4_K_M.gguf --ui-server --max-model-len 32768 --context-cache
 ```
@@ -376,6 +376,7 @@ pip install target/wheels/vllm_rs-*-cp38-abi3-*.whl --force-reinstall
 | `--ui-server`       |  server mode: start the API server and also start the ChatGPT-like web server |
 | `--kv-fraction`       |  control kvcache usage (percentage of remaining gpu memory after model loading) |
 | `--context-cache`   | Enable context caching for multi-turn conversations |
+| `--foce-cache`       | Enable fingerprint-based session detection when `--context-cache` also enabled, no `session_id` needed in this case for context cache |
 
 ### MCP Configuration
 

--- a/example/server.py
+++ b/example/server.py
@@ -36,6 +36,7 @@ def parse_args():
     parser.add_argument("--mcp_config", type=str, default=None)
     parser.add_argument("--mcp_command", type=str, default=None)
     parser.add_argument("--mcp_args", type=str, default=None)
+    parser.add_argument("--force-cache", action="store_true")
 
     return parser.parse_args()
 
@@ -78,6 +79,7 @@ def run_server(args):
         mcp_config=args.mcp_config,
         mcp_command=args.mcp_command,
         mcp_args=args.mcp_args,
+        force_cache=args.force_cache,
     )
 
     engine = Engine(cfg, args.dtype)

--- a/src/core/engine.rs
+++ b/src/core/engine.rs
@@ -1188,10 +1188,11 @@ impl LLMEngine {
         for (param, messages) in params.iter().zip(message_list.iter()) {
             // Fingerprint-based automatic session detection
             let mut param = param.clone();
-            if param.session_id.is_none()
-                && self.econfig.flash_context.unwrap_or(false)
-                && self.econfig.force_cache.unwrap_or(false)
-            {
+            let has_session = param.session_id.is_some();
+            let flash_enabled = self.econfig.flash_context.unwrap_or(false);
+            let force_enabled = self.econfig.force_cache.unwrap_or(false);
+
+            if !has_session && flash_enabled && force_enabled {
                 // Build messages as (role, content) tuples for fingerprint computation
                 let msg_tuples: Vec<(String, String)> = messages
                     .iter()
@@ -1203,6 +1204,10 @@ impl LLMEngine {
                     .fingerprint_manager
                     .find_or_create_session(fingerprint, &msg_tuples);
                 param.session_id = Some(session_id);
+            } else if force_enabled && !flash_enabled {
+                crate::log_warn!(
+                    "force_cache is enabled but flash_context is not - fingerprint detection skipped"
+                );
             }
 
             let (prompt, image_idx) = self.apply_chat_template(&param, messages, false);
@@ -1395,10 +1400,11 @@ impl LLMEngine {
     ) -> Result<(usize, usize, mpsc::Receiver<StreamItem>)> {
         // Fingerprint-based automatic session detection
         let mut params = params.clone();
-        if params.session_id.is_none()
-            && self.econfig.flash_context.unwrap_or(false)
-            && self.econfig.force_cache.unwrap_or(false)
-        {
+        let has_session = params.session_id.is_some();
+        let flash_enabled = self.econfig.flash_context.unwrap_or(false);
+        let force_enabled = self.econfig.force_cache.unwrap_or(false);
+
+        if !has_session && flash_enabled && force_enabled {
             // Build messages as (role, content) tuples for fingerprint computation
             let msg_tuples: Vec<(String, String)> = messages
                 .iter()
@@ -1410,6 +1416,10 @@ impl LLMEngine {
                 .fingerprint_manager
                 .find_or_create_session(fingerprint, &msg_tuples);
             params.session_id = Some(session_id);
+        } else if force_enabled && !flash_enabled {
+            crate::log_warn!(
+                "force_cache is enabled but flash_context is not - fingerprint detection skipped"
+            );
         }
 
         let (prompt, image_idx) = self.apply_chat_template(&params, messages, false);

--- a/src/core/runner.rs
+++ b/src/core/runner.rs
@@ -80,6 +80,8 @@ pub struct ModelRunner {
     seq_tokens: RwLock<HashMap<usize, Vec<u32>>>,
     guidance_states: RwLock<HashMap<usize, GuidanceState>>,
     transfer: Option<Arc<Transfer>>,
+    /// Whether this runner is on the first rank (for logging)
+    is_first_rank: bool,
 }
 
 impl ModelRunner {
@@ -194,6 +196,7 @@ impl ModelRunner {
             seq_tokens: RwLock::new(HashMap::new()),
             guidance_states: RwLock::new(HashMap::new()),
             transfer,
+            is_first_rank: comm.rank() == 0,
         })
     }
 
@@ -702,11 +705,13 @@ impl ModelRunner {
                     });
                 let user_params = &seqs[0].sampling_params;
 
-                // Log thinking parameter whenever user provides it (independent of other sampling config)
-                crate::log_warn!(
-                    "User's thinking preference for reasoning models: {:?}",
-                    user_params.thinking
-                );
+                // Log thinking parameter only from first rank to avoid duplicate logs in multi-GPU
+                if self.is_first_rank && seqs[0].num_cached_tokens == 0 {
+                    crate::log_warn!(
+                        "User's thinking preference for reasoning models: {:?}",
+                        user_params.thinking
+                    );
+                }
 
                 // Determine frequency/presence penalties (user params > generation_cfg)
                 let gen_cfg_freq = self
@@ -724,37 +729,43 @@ impl ModelRunner {
 
                 let sampling = if has_valid_sampling_cfg {
                     let cfg = self.config.generation_cfg.as_ref().unwrap();
-                    crate::log_warn!(
-                        "Using sampling from generation_config: temp={:?}, top_k={:?}, top_p={:?}, freq_penalty={:?}, pres_penalty={:?}",
-                        cfg.temperature,
-                        cfg.top_k,
-                        cfg.top_p,
-                        frequency_penalty,
-                        presence_penalty
-                    );
+                    if self.is_first_rank && seqs[0].num_cached_tokens == 0 {
+                        crate::log_warn!(
+                            "Using sampling from generation_config: temp={:?}, top_k={:?}, top_p={:?}, freq_penalty={:?}, pres_penalty={:?}",
+                            cfg.temperature,
+                            cfg.top_k,
+                            cfg.top_p,
+                            frequency_penalty,
+                            presence_penalty
+                        );
+                    }
                     LogitsProcessor::get_strategy(cfg.temperature, cfg.top_k, cfg.top_p)
                 } else {
                     let has_user_config = matches!(user_params.temperature, Some(t) if t != 0.0 && t != 1.0)
                         && (matches!(user_params.top_k, Some(k) if k > 0)
                             || matches!(user_params.top_p, Some(p) if p != 0.0 && p != 1.0));
                     if has_user_config {
-                        crate::log_warn!(
-                            "Using user's sampling params: temp={:?}, top_k={:?}, top_p={:?}, freq_penalty={:?}, pres_penalty={:?}",
-                            user_params.temperature,
-                            user_params.top_k,
-                            user_params.top_p,
-                            frequency_penalty,
-                            presence_penalty
-                        );
+                        if self.is_first_rank && seqs[0].num_cached_tokens == 0 {
+                            crate::log_warn!(
+                                "Using user's sampling params: temp={:?}, top_k={:?}, top_p={:?}, freq_penalty={:?}, pres_penalty={:?}",
+                                user_params.temperature,
+                                user_params.top_k,
+                                user_params.top_p,
+                                frequency_penalty,
+                                presence_penalty
+                            );
+                        }
                         LogitsProcessor::get_strategy(
                             user_params.temperature,
                             user_params.top_k,
                             user_params.top_p,
                         )
                     } else {
-                        crate::log_warn!(
-                            "No generation_config, using default sampling (temperature=0.7, top_k=32, top_p=0.95)"
-                        );
+                        if self.is_first_rank && seqs[0].num_cached_tokens == 0 {
+                            crate::log_warn!(
+                                "No generation_config, using default sampling (temperature=0.7, top_k=32, top_p=0.95)"
+                            );
+                        }
                         Sampling::TopKThenTopP {
                             k: 32,
                             p: 0.95,

--- a/src/py/mod.rs
+++ b/src/py/mod.rs
@@ -286,7 +286,7 @@ impl EngineConfig {
         generation_cfg=None, seed=None, flash_context = None, fp8_kvcache=None,
         server_mode=None, cpu_mem_fold=None, kv_fraction=None, pd_config=None,
         mcp_command=None, mcp_config=None, mcp_args=None,
-        disable_flash_attn=None))]
+        disable_flash_attn=None, force_cache=None))]
     pub fn new(
         model_id: Option<String>,
         weight_path: Option<String>,
@@ -312,6 +312,7 @@ impl EngineConfig {
         mcp_config: Option<String>,
         mcp_args: Option<String>,
         disable_flash_attn: Option<bool>,
+        force_cache: Option<bool>,
     ) -> Self {
         let mut device_ids = device_ids.unwrap_or_default();
         if device_ids.is_empty() {
@@ -362,6 +363,7 @@ impl EngineConfig {
             mcp_config,
             mcp_args,
             disable_flash_attn,
+            force_cache,
         }
     }
 }

--- a/vllm_rs.pyi
+++ b/vllm_rs.pyi
@@ -73,6 +73,7 @@ class EngineConfig:
     mcp_config: Optional[str]
     mcp_command: Optional[str]
     mcp_args: Optional[str]
+    force_cache: Optional[bool]
 
 @dataclass
 class SamplingParams:

